test/unit/ngsiv2/ngsiService/subscriptions-test.js

```diff
@@ -0,0 +1,351 @@
+/*
+ * Copyright 2016 Telefonica Investigación y Desarrollo, S.A.U
+ *
+ * This file is part of fiware-iotagent-lib
+ *
+ * fiware-iotagent-lib is free software: you can redistribute it and/or
+ * modify it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the License,
+ * or (at your option) any later version.
+ *
+ * fiware-iotagent-lib is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public
+ * License along with fiware-iotagent-lib.
+ * If not, seehttp://www.gnu.org/licenses/.
+ *
+ * For those usages not covered by the GNU Affero General Public License
+ * please contact with::[contacto@tid.es]
+ *
+ * Modified by: Daniel Calvo - ATOS Research & Innovation
+ */
+'use strict';
+
+var iotAgentLib = require('../../../../lib/fiware-iotagent-lib'),
+    utils = require('../../../tools/utils'),
+    should = require('should'),
+    request = require('request'),
+    nock = require('nock'),
+    moment = require('moment'),
+    contextBrokerMock,
+    iotAgentConfig = {
+        logLevel: 'FATAL',
+        contextBroker: {
+            host: '192.168.1.1',
+            port: '1026',
+            ngsiVersion: 'v2'
+        },
+        server: {
+            port: 4041
+        },
+        types: {},
+        service: 'smartGondor',
+        subservice: 'gardens',
+        providerUrl: 'http://smartGondor.com',
+        deviceRegistrationDuration: 'P1M',
+        throttling: 'PT5S'
+    };
+
+describe('Subscription tests', function() {
+    beforeEach(function(done) {
+        var optionsProvision = {
+            url: 'http://localhost:' + iotAgentConfig.server.port + '/iot/devices',
+            method: 'POST',
+            json: utils.readExampleFile('./test/unit/examples/deviceProvisioningRequests/provisionMinimumDevice.json'),
+            headers: {
+                'fiware-service': 'smartGondor',
+                'fiware-servicepath': '/gardens'
+            }
+        };
+
+        nock.cleanAll();
+
+        iotAgentLib.activate(iotAgentConfig, function() {
+            contextBrokerMock = nock('http://192.168.1.1:1026')
+                .matchHeader('fiware-service', 'smartGondor')
+                .matchHeader('fiware-servicepath', '/gardens')
+                .post('/v1/updateContext',
+                    utils.readExampleFile('./test/unit/examples/contextRequests/createMinimumProvisionedDevice.json'))
+                .reply(200,
+                    utils.readExampleFile('./test/unit/examples/contextResponses/createProvisionedDeviceSuccess.json'));
+
+            contextBrokerMock = nock('http://192.168.1.1:1026')
```

Comment:
Not sure how actually mocks in node works... but I'd say that you are asigning the `contextBrokerMock` variable a new value, i.e. a new instance of nock(), thus overriding the value set at L67. I mean, L67-L73 could be removed, isn't it?

Reply:
This does not override the mock, it is just a way to define a new mocking object.

So what it is done is the following:
- In line 67, we declare a mock to provide an appropriate answer to the device provision. In line 73, we just define how is this answer (code 200 and payload the content of createProvisionedDeviceSuccess.json)
- In line 75, we declare another mock for the subscription creation.

These two mocks are called sequentially and you can find plenty of examples in the rest of tests of the IOTAs. In fact, I have just followed what it is already done. ;)

Reply:
Not sure if I'm understanding correctly...

In general, I understant that the following code:
```
contextBrokerMock = A;
contextBrokerMock  = B;
```
is equivalnete to just:

```
contextBrokerMock = B;
```
Given that the second statement overrides the first. I'm not an expect in JS and I could be missing some subtle detail here, but al least this is the way it works in other lenguages I know (Python, C, etc.).

Maybe an "empirical" check could be done... I mean, if you remove the code from L67-L74, does the test still work?

Reply:
I have removed the code L67-L74 and the tests failed:

```
time=2018-01-08T14:53:12.354Z | lvl=DEBUG | corr=9adb0196-432b-4fba-ad94-f19eeeec51dd | trans=9adb0196-432b-4fba-ad94-f19eeeec51dd | op=IoTAgentNGSI.DeviceService | srv=smartGondor | subsrv=/gardens | msg=Creating initial entity in the Context Broker:
 {
    "url": "http://192.168.1.1:1026/v1/updateContext",
    "method": "POST",
    "json": {
        "contextElements": [
            {
                "type": "MicroLights",
                "isPattern": "false",
                "id": "FirstMicroLight",
                "attributes": [
                    {
                        "name": "attr_name",
                        "type": "string",
                        "value": " "
                    }
                ]
            }
        ],
        "updateAction": "APPEND"
    },
    "headers": {
        "fiware-service": "smartGondor",
        "fiware-servicepath": "/gardens",
        "fiware-correlator": "9adb0196-432b-4fba-ad94-f19eeeec51dd"
    }
} | comp=IoTAgent
time=2018-01-08T14:53:12.441Z | lvl=ERROR | corr=9adb0196-432b-4fba-ad94-f19eeeec51dd | trans=9adb0196-432b-4fba-ad94-f19eeeec51dd | op=IoTAgentNGSI.DeviceService | srv=smartGondor | subsrv=/gardens | msg=ORION-001: Connection error creating inital entity in the Context Broker: Error: Nock: No match for request {
  "method": "POST",
  "url": "http://192.168.1.1:1026/v1/updateContext",
  "headers": {
    "fiware-service": "smartGondor",
    "fiware-servicepath": "/gardens",
    "fiware-correlator": "9adb0196-432b-4fba-ad94-f19eeeec51dd",
    "host": "192.168.1.1:1026",
    "accept": "application/json",
    "content-type": "application/json",
    "content-length": 175
  },
  "body": "{\"contextElements\":[{\"type\":\"MicroLights\",\"isPattern\":\"false\",\"id\":\"FirstMicroLight\",\"attributes\":[{\"name\":\"attr_name\",\"type\":\"string\",\"value\":\" \"}]}],\"updateAction\":\"APPEND\"}"
} | comp=IoTAgent

```

As you can see, the test is not able to find the interceptor for the request to provision the device in the CB.

I understand your vision of how this should work because you are thinking about an object assignment but in this case, I guess that each time you call to nock module what you are actually doing is adding a new interceptor to a list. You can read in https://github.com/node-nock/nock#read-this---about-interceptors, that these interceptors are even removed after being used.

Reply:
So you mean that nock module keeps some kind of internal table (I understand that using the URL as key) which is "accumulated" each time nock() is called no mater variable assignation logic? In that case I understand that `contextBrokerMock` variable is irrelevant in this case and probably the current code:

```
contextBrokerMock = nock('http://192.168.1.1:1026')
                                          .someOpsInTheFirstStatement()

contextBrokerMock = nock('http://192.168.1.1:1026')
                                          .someOpsInTheSecondStatement()
```
could be simplified just using 

```
nock('http://192.168.1.1:1026')
        .someOpsInTheFirstStatement()

nock('http://192.168.1.1:1026')
        .someOpsInTheSecondStatement()
```

Is that correct? If I'm right, I think it would be less counter-intuitive that way... having a variable that doesn't follow the usual patterns and that is not revelant at the end is somehow confussing.

Reply:
`contextBrokerMock` is actually used to check that all interceptors have been called. You can check this usage in https://github.com/telefonicaid/iotagent-node-lib/blob/master/test/unit/ngsiService/subscriptions-test.js#L102. If it were not for this, we could modify the code as you propose. Anyway, this kind of use is quite common in the rest of tests of the project, although I have to recognize that was also strange for me because I was also used to work with other languages (C, C++, Java...).

Reply:
Ok. Now its clear. We can go this way.

IMHO is a crazy pattern, but if nock module designers have decided so :) and in addition is a pattern familiar to other people working with nock module, my opinion is subjective. However, maybe some comment would help people not so familiar with nock (as me) not to fall in the same missunderstanding. Something like:

> Note each request interception is declared separatetely with a nock() invocation on the same URL. The assignment is done all the time in the same variable (contextBrokerMock), i.e. new interceptor declarations don't overrides previous ones.

or a similar wording.

Please end this comment thread with either NTC (if you think is ok as it is now, without need extra comment) or with "Fixed in" (if you add a final commit with such comment in the appropiate files)

Btw thanks for the detailed explanation. Now I know how nock() works :)

Reply:
NTC. to close this comment thread... discussion has been diverted to https://github.com/dcalvoalonso/iotagent-node-lib/pull/4#pullrequestreview-87273778)

Reply:
I understand b5f988b overrides this discussion. Thus, we can consider close this comment thread.